botocore/client.py

```diff
@@ -325,6 +246,144 @@ def _api_call(self, *args, **kwargs):
         return _api_call
 
 
+class ClientEndpointBridge(object):
+    """Bridges endpoint data and client creation
+
+    This class handles taking out the relevant arguments from the endpoint
+    resolver and determining which values to use, taking into account any
+    client configuration options and scope configuration options.
+
+    This class also handles determining what, if any, region to use if no
+    explicit region setting is provided. For example, Amazon S3 client will
+    utilize "us-east-1" by default if no region can be resolved."""
+
+    SIGNATURE_PRECEDENCE = ['s3', 's3v4', 'v4', 'v2']
```

Comment:
The preference order is a way for us to check for signature versions that we explicitly have signers for. If a service came out with a signer that we didn't understand, then none of our preferred signers would match and we'd use sigv4. The preferred signers order is checked in a specific order to capture how Botocore favors signature versions. We would use v2 for services that only have v2 support, but only if explicitly stated and only as a last resort. The preference order is SDK specific because other SDKs use the same data source but might prefer v4 over s3 (e.g., PHP and Ruby).